hub/core/chunk_engine.py

```diff
@@ -569,20 +671,55 @@ def numpy(
         enc = self.chunk_id_encoder
 
         for global_sample_index in index.values[0].indices(length):
-            chunks = self.get_chunks_for_sample(global_sample_index)
-
-            if len(chunks) == 1:
-                chunk = chunks[0]
-                idx = enc.translate_index_relative_to_chunks(global_sample_index)
-                sample = chunk.read_sample(idx)
+            chunk_ids = enc[global_sample_index]
+            if not self._is_tiled_sample(global_sample_index):
+                chunk = self.get_chunk_from_chunk_id(chunk_ids[0])
+                enc = self.chunk_id_encoder
+                local_sample_index = enc.translate_index_relative_to_chunks(
+                    global_sample_index
+                )
+                sample = chunk.read_sample(local_sample_index)[
+                    tuple(entry.value for entry in index.values[1:])
+                ]
+            elif len(index.values) == 1:
+                # Tiled sample, all chunks required
+                chunks = self.get_chunks_for_sample(global_sample_index)
+                sample = combine_chunks(chunks, global_sample_index, self.tile_encoder)
             else:
+                # Tiled sample, only some chunks required
                 tile_enc = self.tile_encoder
-                sample = combine_chunks(chunks, global_sample_index, tile_enc)
-            shape = sample.shape
-            check_sample_shape(shape, last_shape, self.key, index, aslist)
+                sample_shape = tile_enc.get_sample_shape(global_sample_index)
+                tile_shape = tile_enc.get_tile_shape(global_sample_index)
+                ordered_tile_ids = np.array(chunk_ids).reshape(
+                    tile_enc.get_tile_layout_shape(global_sample_index)
+                )
+                tiles_index, sample_index = translate_slices(
+                    [v.value for v in index.values[1:]], sample_shape, tile_shape  # type: ignore
+                )
+                required_tile_ids = ordered_tile_ids[tiles_index]
+                tiles = np.vectorize(
+                    lambda chunk_id: self.get_chunk_from_chunk_id(chunk_id).read_sample(
+                        0
+                    ),
+                    otypes=[object],
+                )(required_tile_ids)
+                sample = coalesce_tiles(
+                    tiles, tile_shape, sample_shape, self.tensor_meta.dtype
+                )
+                sample = sample[sample_index]
             samples.append(sample)
-            last_shape = shape
-        return format_read_samples(samples, index, aslist)
```

Comment:
why aren't we using this function anymore?

Reply:
The function can not handle the indexing logic for tiled samples, as the whole sample might not be available. Behavior stays same for non-tile samples, the logic is moved to `numpy()`.